### PR TITLE
fix(mz) app-passwords replacing secrets and client_id. 

### PR DESCRIPTION
**Pull request by joacoc**
### Motivation

Added the `app-password` feature and updated the documentation to reflect the changes. It is the last critical feature of the design document.

### Tips for reviewer

It lets the user create an app password like in the cloud UI. Replaces Frontegg `secret` and `client_id`.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

File: src/mz/Readme.md [hunk 1]
```diff
@@ -58,46 +59,60 @@ The configuration file stores all the available profiles. You can add your own a
...
+
+* Windows: `%UserProfile%\.config\mz\profiles.toml`
 
 Example:
 ```TOML
 ["profiles.production"]
 email = "account@example.com"
-secret = "e1620d58-b3f5-454c-bdc8-6ac07da79fd1"
-client_id = "f599e025-11b3-4fd8-876f-293787dca3c9"
+app-password = "mzp_e1620d58b3f5454cbdc86ac07da79fd1f599e02511b34fd8876f293787dca3c9"
```
Comment (antifuchs) [inline src/mz/Readme.md:76]: I assume this isn't a real one, right?

Comment (antifuchs) [reply src/mz/Readme.md:76]: In either case, I'd like to make it clear that under no circumstances will that password work (but the format is similar):

```suggestion
app-password = "mzp_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
```

File: src/mz/Readme.md [hunk 2]
```diff
@@ -58,46 +59,60 @@ The configuration file stores all the available profiles. You can add your own a
...
 
 ["profiles.staging"]
 email = "account@example.com"
-secret = "d0061af2-f4fd-4b31-a5b2-24efac4df25b"
-client_id = "c01b79a5-6152-45a3-a8b8-2482bc728458"
+app-password = "mzp_c01b79a5615245a3a8b82482bc728458d0061af2f4fd4b31a5b224efac4df25b"
```
Comment (antifuchs) [inline src/mz/Readme.md:80]:
```suggestion
app-password = "mzp_bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
```

File: src/mz/src/main.rs [hunk 3]
```diff
@@ -62,8 +67,8 @@ enum Commands {
         #[clap(short, long)]
         interactive: bool,
     },
-    /// Show commands for interaction with the region
-    Regions(Regions),
+    /// Show commands to interact with regions
+    Region(CRegion),
```
Comment (antifuchs) [inline src/mz/src/main.rs:71]: Can this be an inline struct instead, like the `Shell` below? I think that saves us from adding too many names here.

File: src/mz/src/utils.rs [hunk 4]
```diff
@@ -88,3 +88,39 @@ pub(crate) fn exit_with_fail_message(message: ExitMessage) -> ! {
...
+        None
+    } else {
+        let client_id = layout_uuid(password[4..36].to_string());
+        let secret = layout_uuid(password[36..68].to_string());
+        Some(FronteggAPIToken { client_id, secret })
+    }
+}
+
+/// Format a Materialize password using the client_id and secret
+pub(crate) fn format_password(client_id: String, secret: String) -> String {
```
Comment (antifuchs) [inline src/mz/src/utils.rs:119]: Make this a method on `FronteggAPIToken` instead? That way, you don't have to get the strings to correctly line up.

File: src/mz/src/utils.rs [hunk 5]
```diff
@@ -88,3 +88,39 @@ pub(crate) fn exit_with_fail_message(message: ExitMessage) -> ! {
...
+    ("mzp_".to_owned() + &client_id + &secret).replace('-', "")
+}
+
+/// Turn a profile into a Materialize cloud instance password
+pub(crate) fn password_from_api_token(api_token: FronteggAPIToken) -> String {
```
Comment (antifuchs) [inline src/mz/src/utils.rs:124]: Same, as with `format_password`, I think this should be one single method ideally.

File: src/mz/src/utils.rs [hunk 6]
```diff
@@ -88,3 +88,39 @@ pub(crate) fn exit_with_fail_message(message: ExitMessage) -> ! {
     }
     exit(1);
 }
+
+/// Layouts an UUID.
+/// Frontegg API needs the UUID with the layout.
+fn layout_uuid(uuid: String) -> String {
```
Comment (antifuchs) [inline src/mz/src/utils.rs:94]: It feels kinda weird here to assume that the string is always going to have the wholly-correct format. Can this be structured such that it parses the `Uuid` and then formats it as `uuid.as_simple().to_string()`?

I guess ideally, the `FronteggAPIToken` could parse the mzp_ string into the two UUIDs and validate that they parse right; then everything coming out of it is guaranteed to be valid.

**Review by antifuchs — COMMENTED** [review]
Looks good overall, but passwords handling demands more scrutiny - I'd love to validate the strings there more & left some comments to that effect.

File: src/mz/Readme.md (hunk 1 above)
Comment (joacoc) [reply src/mz/Readme.md:76]: Correct assumption!

File: src/mz/src/main.rs (hunk 3 above)
Comment (joacoc) [reply src/mz/src/main.rs:71]: I revisited [clap docs](https://docs.rs/clap/latest/clap/_derive/index.html#terminology): `struct (arguments) or enum (subcommands).`

Unfortunately, enums can't be anonymous in rust (and inline), which makes me unable to make it the other way :(.

Comment (joacoc) [reply src/mz/src/main.rs:71]: I don't know how but I managed to add a struct as a subcommand using derives.

File: src/mz/src/utils.rs (hunk 6 above)
Comment (joacoc) [reply src/mz/src/utils.rs]: Added the `uuid` crate to do the parsing and checks.

File: src/mz/src/main.rs (hunk 3 above)
Comment (antifuchs) [reply src/mz/src/main.rs]: Hah, we do the thing I suggested on line [73](https://github.com/MaterializeInc/materialize/pull/14970/files/d59a9a63aa47b64288eba7aa178f1e911e2adc1f#diff-bea12c15a2b2c74c54585d6fa33e63d2457564ac96bd634bee6b3af4a57f6aeeR73) (:

Comment (joacoc) [reply src/mz/src/main.rs]: Exactly!

**Comment by joacoc** [comment]
@antifuchs, does it look ready now?

**Review by antifuchs — COMMENTED** [review]
Looks good (in retrospect)! Thanks for pushing this over the finish line!

**Pull request merged**